sunxspex/sunxspex_fitting/fitter.py

```diff
@@ -0,0 +1,4931 @@
+            and work nroally when loaded back in.
+
+    Returns
+    -------
+    None.
+
+    Example
+    -------
+    from fitter import SunXspex, defined_photon_models, add_photon_model, add_var
+
+    # the user variable that might be too costly to run every function call or too hard to hard code
+    some_user_var = something_complicated
+
+    # Define gaussian model (doesn't have to be a lambda function), but include some user variable outside the scope of the model
+    gauss = lambda a, b, c, energies=None: a * np.exp(-((np.mean(energies, axis=1)-b)**2/(2*c**2))) * some_user_var
+
+    # add user variable
+    add_var(some_user_var=some_user_var)
+
+    # Add the gaussian model to fitter.py namespace
+    add_photon_model(gauss)
+
+    # Now can use it in fitting with string defined model. Will be plotted separately to the total model
+    Sx = SunXspex(pha_file=[...])
+    Sx.model = "gauss+gauss"
+    Sx.fit()
+    Sx.plot()
+    """
+    for k,i in user_kwarg.items():
+        if k in DYNAMIC_VARS and not overwrite:
+            vb = f"Variable {k} already exists. Please set `overwrite=True`, delete this with `del_var({k})`,\nor use a different variable name."
+        elif not k in DYNAMIC_VARS and k in globals():
+            vb = f"Argument name {k} already exists **in globals** and is not a good idea to overwrite. Please use a different variable name."
+        else:
+            DYNAMIC_VARS.update({k:i})
+            globals().update({k:i})
+            vb = f"Variable {k} added."
+        if not quiet:
+            print(vb)
+
+def del_var(*user_arg_name):
+    """ Remove user defined variables that have been added via `add_var`.
+
+    Parameters
+    ----------
+    *user_arg_name : str
+            Name(s) of the variable(s) to be removed.
+
+    Returns
+    -------
+    None.
+
+    Example
+    -------
+    from fitter import add_var, del_var
+
+    # the user variable that might be too costly to run every function call or too hard to hard code
+    some_user_var = something_complicated
+
+    # add user variable
+    add_var(some_user_var=some_user_var)
+
+    # realise the variable should be there or want it gome to update it
+    del_var("some_user_var")
+    """
+    _removed, _not_removed = [], []
+    for uan in user_arg_name:
+        if uan in DYNAMIC_VARS:
+            del globals()[uan], DYNAMIC_VARS[uan]
+            _removed.append(uan)
+        else:
+            _not_removed.append(uan)
+    _rmstr, spc = (f"Variables {_removed} were removed.", "\n") if len(_removed)>0 else ("", "")
+    _nrmstr, spc  = (f"Variables {_not_removed} are not ones added by user and so were not removed.", spc) if len(_not_removed)>0 else ("", "")
+    print(_rmstr, _nrmstr, sep=spc)
+
+# Easily access log-likelihood/fit-stat methods from the one place, if SunXpsex class inherits this then data is duplicated
+LL_CLASS = LogLikelihoods()
+
+class SunXspex(LoadSpec):
```

Comment:
Absolutely. I believe this should be able to be split up quite a bit. This class did continue to keep growing when I just kept thinking "another method won't hurt" or "oh, it could do this too". I wouldn't have constructed a huge class like this if I could go back so some splitting up is required.